src/reactor.rs

```diff
@@ -0,0 +1,741 @@
+use crate::{Consensus, Error, Height, Keychain, Metadata, Moderator, Round, Step, Vote};
+
+use fuel_crypto::PublicKey;
+use fuel_types::Bytes32;
+use time::OffsetDateTime;
+
+use alloc::vec::Vec;
+use core::time::Duration;
+
+mod config;
+mod message;
+
+pub use config::Config;
+pub use message::{Event, Message, Notification, Request, Response};
+
+/// State machine of the consensus
+#[derive(Debug)]
+pub struct Reactor {
+    capacity: usize,
+    consensus: u128,
+    genesis: OffsetDateTime,
+    metadata: Metadata,
+    timeout: Duration,
+    should_quit: bool,
+}
+
+impl Default for Reactor {
+    fn default() -> Self {
+        Self::new(Default::default())
+    }
+}
+
+impl Reactor {
+    /// Create a new reactor with the provided arguments
+    pub fn new(config: Config) -> Self {
+        let Config {
+            capacity,
+            consensus,
+            genesis,
+            timeout,
+            ..
+        } = config;
+
+        let metadata = Default::default();
+        let should_quit = false;
+
+        Self {
+            capacity,
+            consensus,
+            genesis,
+            metadata,
+            timeout,
+            should_quit,
+        }
+    }
+
+    /// Flag on whether should quit
+    pub const fn should_quit(&self) -> bool {
+        self.should_quit
+    }
+
+    /// Current block height.
+    pub const fn height(&self) -> Height {
+        self.metadata.committed_height().wrapping_add(1)
+    }
+
+    /// Current height round
+    pub fn round(&self, now: OffsetDateTime) -> Round {
+        let elapsed = now - self.genesis;
```

Comment:
This deviates from tendermint-style algorithms and could lead to rounds being arbitrarily skipped. This also doesn't allow for the eventuality guarantees of tendermint's global stabilization time. Namely, it eliminates the following properties that allow GST to occur:

> The timeouts prevent the algorithm from blocking and waiting forever for some condition to be true, ensure that processes continuously transition between rounds, and guarantee that eventually (after GST) communication between correct processes is timely and reliable so they can decide. The last role is achieved by increasing the timeouts with every new round r, i.e,`timeoutX(r) = initT imeoutX +r ∗ timeoutDelta;` they are reset for every new height (consensus instance).

[tendermint paper](https://arxiv.org/pdf/1807.04938.pdf#page=5)

Where timeoutX could be timeoutPropose, timeoutPrevote or timeoutPrecommit.

This means we should only increment the round _after_ properly cycling through the appropriate steps and phases, rather than just taking the time since the last commit and dividing it by some constant round value. Otherwise, we won't be able to increasing the timeout each round.

Additionally, we may want to consider a timeout vote in the future which would allow for ddos-resistant leader selection strategies.

Reply:
In a way this acts similar to an exponential backoff retry strategy that resets on every new block height.

Reply:
This ends up being a matter of what level of granularity we want to support. As you correctly pointed out, we now perform timeout per round instead of per step, but then this wouldn't violate any of the guarantees of the protocol.

Both approaches (per round or per step) would simply divide the spent time since last checkpoint by a constant. I opened one issue here to discuss that trade-off and not further increase the scope of this PR: https://github.com/FuelLabs/fuel-bft/issues/13

Reply:
The main point I'm trying to make here is that the timeout increases each successive round, not so much that there's a granular timeout each step.

Reply:
The other strange thing about using a constant time per round this way is that if the network is offline for some period of time, the next round would suddenly jump to some arbitrary height.

Reply:
The other scenario shouldn't happen. There is even a test case covering that:

https://github.com/FuelLabs/fuel-bft/blob/bb6b7dc5ad66905ad9d15c8a1fe35329bc37e770/tests/vectors/leader.yaml#L46-L54

Reply:
Let's implement the delta in a follow-up PR? This PR already contains many different issues